contributors/design-proposals/network/grpc-cni-plugin.md

```diff
@@ -0,0 +1,268 @@
+
+A daemon process will be offered to capture all the RPCs from kubelet and handle it or call CNI plugin binaries.
+
+## Design Details
+
+#### Kubernetes In-tree remote CNI network plugin
+
+Build a new `RemoteCNINetworkPlugin` object, which will **realize all the interface `NetworkPlugin` required**, this will enable kubelet use the new network plugin **just like CNI and kubenet**. And once the CNI plugin DaemonSet is stable and function as the the old in-tree cni network plugin, the old one will be deprecated.
+
+The `RemoteCNINetworkPlugin` object will include a gRPC client named `cniCLient`, the kubelet invokes for `RemoteCNINetworkPlugin` will be convert to corresponding RPCs calls by `cniCLient`.
+
+```go
+// RemoteCNINetworkPlugin is a GRPC implementation of NetworkPlugin
+type RemoteCNINetworkPlugin struct {
+
+
+  // host is an interface that plugin can use to access the kubelet
+  // plugin will use host to get network namespace information
+  // of sandbox
+  host      network.Host
+
+  // cniClient is the client for CNIDriver service
+  // plugin will convert kubelet call to invokes of this client
+  cniClient cniClient
+}
+
+func (plugin *RemoteCNINetworkPlugin) SetUpPod{
+  // *1 Get sandbox network namespace
+  plugin.host.GetNetNS
+
+  // *2 Get portMap
+  plugin.host.GetPodPortMapping
+
+  // *3 build runtime config
+
+  // *4 Call cniClient
+  // cniClient.setUpPod will call DaemonSet CNI driver
+  // to add CNI network
+  plugin.cniClient.setUpPod
+}
+
+func (plugin *RemoteCNINetworkPlugin) TearDownPod{
+  // *1 Get sandbox network namespace
+  plugiun.host.GetNetNS
+
+  // *2 Get portMap
+  plugin.host.GetPodPortMapping
+
+  // *3 build runtime config
+
+  // *4 Call cniClient
+  // cniClient.tearDownPod will call DaemonSet CNI driver
+  // to delete CNI network
+  plugin.cniClient.tearDownPod
+}
+
+func (plugin *RemoteCNINetworkPlugin) Status{
+  // call cniClient
+  // cniCLient.cniStatus will call DaemonSet CNI driver
+  // to sync CNI network config and check whether
+  // CNI driver initialized
+  plugin.cniClient.cniStatus
+}
+```
+
+#### DaemonSet CNI driver program
+
+We will build a new repo for out-tree CNI driver, maybe named `kubernetes-cni`, which will run as a DeamonSet program,  and offer a gRPC server that in-tree network plugin can invoke.  This CNI driver will acts as a bridge between third-party CNI plugin binaries and kubelet.
+
+And the DaemonSet CNI driver will use a configmap as the network config file, so don’t have to access the root filesystem of the node or master.
+
+```go
+type cniDriver struct {
+  // cni.NetworkConfig is the network config specified by
+  // containernetworking/cni
+  // CNI driver will build this config according to network
+  // information from configmap

```

Comment:
If this is coming from a config map, does it mean we'll require this CNI config to be the same across all nodes? What about node-specific configuration? I think this would break a number of plugins (e.g. host-local) where per-node config is passed to the plugin, or it would require one configmap per-node (super ick).

Another example is Calico, where it supports various configuration options in the CNI network config file that some users want to be able to set on a per-node basis. I think requiring that a configmap be used to distribute the CNI network configuration might be too strict. Certainly that is one way it can be done, but in Calico for example we often use a network config _template_ in a ConfigMap, which each node can populate with its own node-specific data.